### PR TITLE
Install whl in Dockerfile

### LINKED ISSUE #70: Stop cloning repo when building image
`pip install` in `Dockerfile` once wheel releases are setup instead of cloning the repo while building the image

### PR DESCRIPTION
**Pull request by tedhtchang**
### Description of the change
- Change fms-hf-tuning python lib is installed in the Dockerfile
- Add GH action to test image build

### Related issue number
May close #70 

### How to verify the PR
```
git clone --single-branch --branch whl-install-dockerfile https://github.com/tedhtchang/fms-hf-tuning.git
cd fms-hf-tuning
docker build --progress=plain -t fms-hf-tuning:dev . -f build/Dockerfile
# Download whl from PyPi instead build
docker build --build-arg WHEEL_VERSION=0.0.1 --progress=plain -t fms-hf-tuning:0.0.1 . -f build/Dockerfile
```
How to test this image in a KinD cluster
Simply
```bash
curl https://gist.githubusercontent.com/tedhtchang/34335240230d64c6e2ecb5b78d8a8cb0/raw/6c568c7fdc82ed9e344474da5278ec43a6fc788b/kind-e2e-sft-mini.sh | bash
```

### Was the PR tested

- [ x] I have added >=1 unit test(s) for every new method I have added.
- [ x] I have ensured all unit tests pass

### CODE REVIEW

**Comment by jbusche** [comment]
Hey @tedhtchang 
I gave this a try.  It looks like the branch name (--branch deprecate-requirements) might be wrong.  I tried these steps on my OC 4.14.17 cluster:
```
git clone --single-branch --branch whl-install-dockerfile https://github.com/tedhtchang/fms-hf-tuning.git
cd fms-hf-tuning
docker build --progress=plain -t fms-hf-tuning:bad . -f build/Dockerfile
```
Then I pushed it to my local registry:
```
podman login -u kubeadmin -p $(oc whoami -t) $(oc registry info) --tls-verify=false
podman tag localhost/fms-hf-tuning:bad $(oc registry info)/opendatahub/fms-hf-tuning:bad
podman push  --tls-verify=false $(oc registry info)/opendatahub/fms-hf-tuning:ba
```
Lastly, I tried your test script with it:
```
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-config
data:
  config.json: |
    {
      "accelerate_launch_args": {
        "num_machines": 1,
        "num_processes": 2
      },
      "model_name_or_path": "bigscience/bloom-560m",
      "training_data_path": "/etc/config/twitter_complaints_small.json",
      "output_dir": "/tmp/out",
      "num_train_epochs": 1.0,
      "per_device_train_batch_size": 4,
      "per_device_eval_batch_size": 4,
      "gradient_accumulation_steps": 4,
      "evaluation_strategy": "no",
      "save_strategy": "epoch",
      "learning_rate": 1e-5,
      "weight_decay": 0.0,
      "lr_scheduler_type": "cosine",
      "logging_steps": 1.0,
      "packing": false,
      "include_tokens_per_second": true,
      "response_template": "\n### Label:",
      "dataset_text_field": "output",
      "use_flash_attn": false,
      "torch_dtype": "float32",
      "peft_method": "pt",
      "tokenizer_name_or_path": "bigscience/bloom"
    }
  twitter_complaints_small.json: |
    {"Tweet text":"@HMRCcustomers No this is my first job","ID":0,"Label":2,"text_label":"no complaint","output":"### Text: @HMRCcustomers No this is my first job\n\n### Label: no complaint"}
    {"Tweet text":"@KristaMariePark Thank you for your interest! If you decide to cancel, you can call Customer Care at 1-800-NYTIMES.","ID":1,"Label":2,"text_label":"no complaint","output":"### Text: @KristaMariePark Thank you for your interest! If you decide to cancel, you can call Customer Care at 1-800-NYTIMES.\n\n### Label: no complaint"}
    {"Tweet text":"If I can't get my 3rd pair of @beatsbydre powerbeats to work today I'm doneski man. This is a slap in my balls. Your next @Bose @BoseService","ID":2,"Label":1,"text_label":"complaint","output":"### Text: If I can't get my 3rd pair of @beatsbydre powerbeats to work today I'm doneski man. This is a slap in my balls. Your next @Bose @BoseService\n\n### Label: complaint"}
    {"Tweet text":"@EE On Rosneath Arial having good upload and download speeds but terrible latency 200ms. Why is this.","ID":3,"Label":1,"text_label":"complaint","output":"### Text: @EE On Rosneath Arial having good upload and download speeds but terrible latency 200ms. Why is this.\n\n### Label: complaint"}
    {"Tweet text":"Couples wallpaper, so cute. :) #BrothersAtHome","ID":4,"Label":2,"text_label":"no complaint","output":"### Text: Couples wallpaper, so cute. :) #BrothersAtHome\n\n### Label: no complaint"}
    {"Tweet text":"@mckelldogs This might just be me, but-- eyedrops? Artificial tears are so useful when you're sleep-deprived and sp\u2026 https:\/\/t.co\/WRtNsokblG","ID":5,"Label":2,"text_label":"no complaint","output":"### Text: @mckelldogs This might just be me, but-- eyedrops? Artificial tears are so useful when you're sleep-deprived and sp\u2026 https:\/\/t.co\/WRtNsokblG\n\n### Label: no complaint"}
    {"Tweet text":"@Yelp can we get the exact calculations for a business rating (for example if its 4 stars but actually 4.2) or do we use a 3rd party site?","ID":6,"Label":2,"text_label":"no complaint","output":"### Text: @Yelp can we get the exact calculations for a business rating (for example if its 4 stars but actually 4.2) or do we use a 3rd party site?\n\n### Label: no complaint"}
    {"Tweet text":"@nationalgridus I have no water and the bill is current and paid. Can you do something about this?","ID":7,"Label":1,"text_label":"complaint","output":"### Text: @nationalgridus I have no water and the bill is current and paid. Can you do something about this?\n\n### Label: complaint"}
    {"Tweet text":"Never shopping at @MACcosmetics again. Every time I go in there, their employees are super rude\/condescending. I'll take my $$ to @Sephora","ID":8,"Label":1,"text_label":"complaint","output":"### Text: Never shopping at @MACcosmetics again. Every time I go in there, their employees are super rude\/condescending. I'll take my $$ to @Sephora\n\n### Label: complaint"}
    {"Tweet text":"@JenniferTilly Merry Christmas to as well. You get more stunning every year \ufffd\ufffd","ID":9,"Label":2,"text_label":"no complaint","output":"### Text: @JenniferTilly Merry Christmas to as well. You get more stunning every year \ufffd\ufffd\n\n### Label: no complaint"}
---
apiVersion: "kubeflow.org/v1"
kind: PyTorchJob
metadata:
  name: ted-kfto-sft
  labels:
    kueue.x-k8s.io/queue-name: lq-trainer
spec:
  pytorchReplicaSpecs:
    Master:
      replicas: 1
      restartPolicy: Never # Do not restart the pod on failure. If you do set it to OnFailure, be sure to also set backoffLimit
      template:
        spec:
          containers:
            - name: pytorch
              # This is the temp location util image is officially released
              image: image-registry.openshift-image-registry.svc:5000/opendatahub/fms-hf-tuning:bad
              #image: quay.io/jbusche/fms-hf-tuning:issue758-1
              imagePullPolicy: IfNotPresent
              command:
                - "python"
                - "/app/accelerate_launch.py"
              env:
                - name: SFT_TRAINER_CONFIG_JSON_PATH
                  value: /etc/config/config.json
              volumeMounts:
              - name: config-volume
                mountPath: /etc/config
          volumes:
          - name: config-volume
            configMap:
              name: my-config
              items:
              - key: config.json
                path: config.json
              - key: twitter_complaints_small.json
                path: twitter_complaints_small.json
EOF
```
And it looked good:
```
oc get pods,pytorchjobs
NAME                        READY   STATUS      RESTARTS   AGE
pod/ted-kfto-sft-master-0   0/1     Completed   0          7m49s

NAME                                   STATE       AGE
pytorchjob.kubeflow.org/ted-kfto-sft   Succeeded   7m49s
```

**Review by jbusche — APPROVED** [review]
LGTM

**Review by alex-jw-brooks — COMMENTED** [review]
Thanks @tedhtchang - this looks reasonable to me. To clarify, the intention of the GitHub workflow being added is to just ensure that the Dockerfile is buildable, and not to actually test the wheel-based install (e.g., running unit tests in an environment where the package was installed & the source was removed from the python path), is that correct?

**Comment by tedhtchang** [comment]
Thanks @jbusche @alex-jw-brooks  
> (--branch deprecate-requirements) might be wrong.
Updated Thanks

> Thanks @tedhtchang - this looks reasonable to me. To clarify, the intention of the GitHub workflow being added is to just ensure that the Dockerfile is buildable, and not to actually test the wheel-based install (e.g., running unit tests in an environment where the package was installed & the source was removed from the python path), is that correct?

That's correct. The additional check verifies the image is buildable. The whl based install and test is covered by [these tests](https://github.com/foundation-model-stack/fms-hf-tuning/actions/runs/8729488664/job/23951521549?pr=126#step:4:15).

File: build/Dockerfile [hunk 1]
```diff
@@ -12,6 +12,24 @@
...
+    python -m pip install --upgrade pip && \
+    python -m pip install build
+COPY tuning tuning
+RUN --mount=source=.git,target=.git,type=bind \
+    --mount=type=bind,source=pyproject.toml,target=pyproject.toml \
+    # build wheel if wheel version is empty else download the wheel from PyPi
+    if [[ -z "${WHEEL_VERSION}" ]]; \
+    then python -m build --wheel --outdir /tmp; \
+    else pip download fms-hf-tuning==${WHEEL_VERSION} --dest /tmp --only-binary=:all: --no-deps; \
+    fi && \
```
Comment (tedhtchang) [inline build/Dockerfile:29]: @alex-jw-brooks @anhuong @jbusche This change optionally allows install versioned whl. For example, you can try:
```
docker build --build-arg WHEEL_VERSION=0.0.1 --progress=plain -t fms-hf-tuning:0.0.1 . -f build/Dockerfile
```

**Review by sutaakar — APPROVED** [review]

File: build/Dockerfile [hunk 2]
```diff
@@ -12,6 +12,24 @@
...
+USER root
+RUN --mount=type=cache,target=/root/.cache/pip \
+    python -m pip install --upgrade pip && \
+    python -m pip install build
+COPY tuning tuning
+RUN --mount=source=.git,target=.git,type=bind \
+    --mount=type=bind,source=pyproject.toml,target=pyproject.toml \
+    # build wheel if wheel version is empty else download the wheel from PyPi
+    if [[ -z "${WHEEL_VERSION}" ]]; \
+    then python -m build --wheel --outdir /tmp; \
```
Comment (sutaakar) [inline build/Dockerfile:27]: When I try to build the image using `docker build --progress=plain -t fms-hf-tuning:0.0.1 . -f build/Dockerfile` I get an error:
```
[1/2] STEP 6/6: RUN --mount=source=.git,target=.git,type=bind     --mount=type=bind,source=pyproject.toml,target=pyproject.toml     if [[ -z "${WHEEL_VERSION}" ]];     then python -m build --wheel --outdir /tmp;     else pip download fms-hf-tuning==${WHEEL_VERSION} --dest /tmp --only-binary=:all: --no-deps;     fi &&     ls /tmp/*.whl >/tmp/bdist_name
* Creating isolated environment: venv+pip...
ERROR Source /opt/app-root/src does not appear to be a Python project: no pyproject.toml or setup.py
Error: building at STEP "RUN --mount=source=.git,target=.git,type=bind --mount=type=bind,source=pyproject.toml,target=pyproject.toml if [[ -z "${WHEEL_VERSION}" ]];     then python -m build --wheel --outdir /tmp;     else pip download fms-hf-tuning==${WHEEL_VERSION} --dest /tmp --only-binary=:all: --no-deps;     fi &&     ls /tmp/*.whl >/tmp/bdist_name": while running runtime: exit status 1
```

Comment (sutaakar) [reply build/Dockerfile:27]: I suppose the error is caused by using Podman instead of Docker

**Comment by jbusche** [comment]
I tried again @tedhtchang on my OC 4.14.17 cluster and the build of the dev image looked good.  I installed and ran it and it ran fine:
```
Every 2.0s: oc get pods,pytorchjobs                                                           api.ted414.cp.fyre.ibm.com: Wed Apr 24 13:46:52 2024

NAME                        READY   STATUS	RESTARTS   AGE
pod/ted-kfto-sft-master-0   0/1     Completed   0          4m25s

NAME                                   STATE	   AGE
pytorchjob.kubeflow.org/ted-kfto-sft   Succeeded   4m25s
```
I also tried @sutaakar with an Ubuntu 22.04.4 image with docker installed `snap install docker` and it seemed to build the image fine:
```
docker build --progress=plain -t fms-hf-tuning:dev . -f build/Dockerfile
....
#28 exporting to image
#28 exporting layers
#28 exporting layers 94.7s done
#28 writing image sha256:7adac34a6ce86d33410ab970255355bbcee59586f6856e7f502d59118b092e59 0.0s done
#28 naming to docker.io/library/fms-hf-tuning:dev 0.1s done
#28 DONE 94.9s

docker images 
REPOSITORY      TAG       IMAGE ID       CREATED             SIZE
fms-hf-tuning   dev       7adac34a6ce8   About an hour ago   12.9GB
```
I also tried the WHEEL version install on both Red Hat and ubuntu (with Docker) and it built OK for me.
```
docker build --build-arg WHEEL_VERSION=0.0.1 --progress=plain -t fms-hf-tuning:0.0.1 . -f build/Dockerfile
```
But when I try to run it, it doesn't work:
```
Traceback (most recent call last):
  File "/app/launch_training.py", line 118, in <module>
    main()
  File "/app/launch_training.py", line 111, in main
    training_args.output_dir, sft_trainer.TRAINING_LOGS_FILENAME
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'tuning.sft_trainer' has no attribute 'TRAINING_LOGS_FILENAME'
```
Maybe the PiPY 0.0.1 version is too old/out of date?

File: build/Dockerfile (hunk 2 above)
Comment (tedhtchang) [reply build/Dockerfile:27]: hmm the build command worked for me on a RHEL9.2 x86 VM
```
podman build --build-arg WHEEL_VERSION=0.0.1 --progress=plain -t fms-hf-tuning:0.0.1 . -f build/Dockerfile
```
I wonder podman machine failed to mount the pyproject.toml from your local host file system.  If your local build uses `podman machine` on MacOS Intel. You should mount your workspace into the podman machine first
For example:
```
podman rm
podman machine init --cpus 6 --memory 8196 -v /Users/sutaakar:/Users/sutaakar --rootful --now
```

Comment (sutaakar) [reply build/Dockerfile:27]: Just an idea, what about using:
```
COPY .git .git
COPY pyproject.toml pyproject.toml
# build wheel if wheel version is empty else download the wheel from PyPi
RUN if [[ -z "${WHEEL_VERSION}" ]]; \
    then python -m build --wheel --outdir /tmp; \
    else pip download fms-hf-tuning==${WHEEL_VERSION} --dest /tmp --only-binary=:all: --no-deps; \
    fi && \
    ls /tmp/*.whl >/tmp/bdist_name
```
This way there won't be any issue with mounting files and should work on all platforms. It won't affect result image the files are copied on builder image only.

Comment (tedhtchang) [reply build/Dockerfile:27]: Great idea. Thanks.

**Review by sutaakar — APPROVED** [review]

**Comment by jbusche** [comment]
Hi @tedhtchang, I tried this on my cluster both as a normal build as well as the wheel build, and both worked great.  I think It's good to go.

**Review by jbusche — APPROVED** [review]
Tried it on my 4.14.17 OpenShift Cluster, looked good.

**Review by anhuong — APPROVED** [review]
LGTM, thanks all!

**Pull request merged**